src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -1014,11 +1017,15 @@ void ReadFromMergeTree::setQueryInfoInputOrderInfo(InputOrderInfoPtr order_info)
 
 ReadFromMergeTree::AnalysisResult ReadFromMergeTree::getAnalysisResult() const
 {
-    auto result_ptr = analyzed_result_ptr ? analyzed_result_ptr : selectRangesToRead(prepared_parts);
-    if (std::holds_alternative<std::exception_ptr>(result_ptr->result))
-        std::rethrow_exception(std::get<std::exception_ptr>(result_ptr->result));
+    if (std::holds_alternative<std::exception_ptr>(analyzed_result_ptr->result))
+        std::rethrow_exception(std::move(std::get<std::exception_ptr>(analyzed_result_ptr->result)));
 
-    return std::get<ReadFromMergeTree::AnalysisResult>(result_ptr->result);
+    return std::get<ReadFromMergeTree::AnalysisResult>(analyzed_result_ptr->result);
+}
+
+const ReadFromMergeTree::AnalysisResult & ReadFromMergeTree::getAnalysisResult() const
+{
+    return const_cast<ReadFromMergeTree *>(this)->getAnalysisResult();
```

Comment:
This makes code less readable.

Reply:
That's the only legitimate use of const_cast as far as I know, to avoid code duplication when overriding const/non-const methods. Similar examples also exist in book C++ Primer.

Reply:
Looks like both versions are const method now, but one return value and one return const ref. The 1st version is a mistake, should be non-const and return `AnalysisResult &`

Reply:
I found that the non-const version of this method currently not used, so I just remove it @alexey-milovidov